File: 2023/UICPC Regional/Round #15/solutions/flowfinder/per.py
n = int(raw_input())
par = [0] + map(int, raw_input().split())
flow = map(int, raw_input().split())

children = [[] for _ in range(n)]
for i in range(1, n):
    par[i] -= 1
    children[par[i]].append(i)

order = [0]
for i in range(n):
    u = order[i]
    order.extend([v for v in children[u]])

minflow = [1]*n
for u in reversed(order):
    minflow[u] = max(1, flow[u], sum(minflow[v] for v in children[u]))

c_minrem = [sum(minflow[v] for v in children[u] if flow[v] <= 0) for u in range(n)]
c_unknown = [sum(1 for v in children[u] if flow[v] <= 0) for u in range(n)]
c_sum = [sum(flow[v] for v in children[u] if flow[v] > 0) for u in range(n)]

Q = []
def assign(u, f):
    flow[u] = f
    if u:
        c_unknown[par[u]] -= 1
        c_minrem[par[u]] -= minflow[u]
        c_sum[par[u]] += flow[u]
    Q.append(u)
def check(u):
    if flow[u] <= 0 and c_unknown[u] == 0 and children[u]:
        assign(u, c_sum[u])
    if (flow[u] > 0 and c_unknown[u] > 0 and
          (c_unknown[u] == 1 or flow[u] - c_sum[u] == c_minrem[u])):
        for v in children[u]:
            if flow[v] <= 0:
                assign(v, (flow[u] - c_sum[u]) * minflow[v] / c_minrem[u])
                
for u in range(n): check(u)
while Q:
    u = Q.pop()
    check(u)
    if u > 0: check(par[u])

if any(c_unknown[u] or flow[u] <= 0 or children[u] and c_sum[u] != flow[u] for u in range(n)):
    print 'impossible'
else:
    print '\n'.join(map(str, flow))

File: 2023/UICPC Regional/Round #15/solutions/flowfinder/per.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

int main(void) {
    int n;
    scanf("%d", &n);
    vector<int> par(n);
    vector<ll> flow(n);
    for (int i = 1; i < n; ++i) scanf("%d", &par[i]), --par[i];
    for (auto &f: flow) scanf("%lld", &f);

    vector<vector<int>> children(n);
    for (int i = 1; i < n; ++i) 
        children[par[i]].push_back(i);

    vector<ll> minflow(n, 1);
    function<ll(int)> find_min = [&](int u) -> ll {
        ll c_sum = 0;
        for (int v: children[u])
            c_sum += find_min(v);
        return minflow[u] = max(1LL, max(flow[u], c_sum));
    };
    find_min(0);

    vector<ll> c_minrem(n), c_unknown(n), c_sum(n);
    for (int u = 0; u < n; ++u)
        for (int v: children[u]) {
            if (flow[v] <= 0) {
                c_minrem[u] += minflow[v];
                ++c_unknown[u];
            } else
                c_sum[u] += flow[v];
        }
    
    vector<int> Q;
    auto assign = [&](int u, ll f) {
        flow[u] = f;
        if (u) {
            --c_unknown[par[u]];
            c_minrem[par[u]] -= minflow[u];
            c_sum[par[u]] += flow[u];
        }
        Q.push_back(u);
    };
    
    auto check = [&](int u) {
        if (flow[u] <= 0 && c_unknown[u] == 0 && !children[u].empty())
            assign(u, c_sum[u]);
        if (flow[u] > 0 && c_unknown[u] > 0 &&
            (c_unknown[u] == 1 || flow[u] - c_sum[u] == c_minrem[u])) {
            for (int v: children[u])
                if (flow[v] <= 0)
                    assign(v, (flow[u] - c_sum[u]) * minflow[v] / c_minrem[u]);
        }
    };
    for (int u = 0; u < n; ++u) check(u);
    while (!Q.empty()) {
        int u = Q.back(); Q.pop_back();
        check(u);
        if (u > 0) check(par[u]);
    }

    bool impossible = false;
    for (int u = 0; !impossible && u < n; ++u)
        impossible |= c_unknown[u] || flow[u] <= 0 || !children[u].empty() && c_sum[u] != flow[u];
    
    if (impossible)
        printf("impossible\n");
    else
        for (int u = 0; u < n; ++u)
            printf("%lld\n", flow[u]);
}
